pkg/workflow/controllers/new_node.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+)
+
+var (
+	isController       = true
+	blockOwnerDeletion = true
+	ApiVersion         = v1alpha1.GroupVersion.String()
+	KindWorkflow       = "Workflow"
+	KindWorkflowNode   = "WorkflowNode"
+)
+
+const LabelControlledBy = "chaos-mesh.org/controlled-by"
+
+// renderNodesByTemplates will render the nodes one by one, will setup owner by given parent. If parent is nil, it will use workflow as its owner.
+func renderNodesByTemplates(workflow *v1alpha1.Workflow, parent *v1alpha1.WorkflowNode, templates ...string) ([]*v1alpha1.WorkflowNode, error) {
+	templateNameSet := make(map[string]v1alpha1.Template)
+	for _, template := range workflow.Spec.Templates {
+		templateNameSet[template.Name] = template
+	}
+	var result []*v1alpha1.WorkflowNode
+	for _, name := range templates {
+		if template, ok := templateNameSet[name]; ok {
+
+			now := metav1.NewTime(time.Now())
+			var deadline *metav1.Time = nil
+
+			if template.Duration != nil {
+				duration, err := time.ParseDuration(*template.Duration)
+				if err != nil {
+					// TODO: logger
+					return nil, err
+				}
+				result := metav1.NewTime(now.DeepCopy().Add(duration))
```

Comment:
how about renaming the `result`, it is duplicate with L42

Reply:
fixed in 187a713